### PR TITLE
Diagonal traversal of Binary Tree

### PR DESCRIPTION
**Pull request by utkarsh006**
## Related Issue
- Title Diagonal Traversal #1296

## Proposed Changes
- I have added the following things: -
- Problem Statement
- Examples
- Algorithm
- Code
- Time Complexity & Space Complexity

## Additional Info
[LINK OF THE PROBLEM STATEMENT](https://practice.geeksforgeeks.org/problems/diagonal-traversal-of-binary-tree/1)
## Checklist

- [x] ✅ My code follows the code style of this project.
- [x] 📝 My change requires a change to the README.
- [x] 🎀 I have updated the README accordingly.
- [x] 👀 I have read the [**CONTRIBUTING**](https://github.com/Lakhankumawat/LearnCPP/blob/main/Contributing.md) document.
- [x] ✨ I have added tests to cover my changes.
- [x] 🚩 All new and existing tests passed.
- [x] 🌟 ed the repo

We were supposed to display the diagonal traversal of the binary tree.

I have added the Problem Statement, Examples, an Algorithm on how to approach this problem, a commented Code, and written the Time Complexity and Space Complexity. And finally updated the README.md.

This change helps to solve a very famous problem of Binary Trees i.e. Diagonal Traversal of Binary Tree and this problem helps in understanding the concept of Trees in depth.

I coded this problem in VS Code and tested the given problem on the test cases in the range: 1 <= Number of nodes<= 105,
1 <= Data of a node<= 105. And it is working fine for them.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Output Screenshots
| Screenshot #1      | Screenshot #2  |
| ----------- | ----------- |
| Input : [1,2,3,4,null,5,6,null,null,7,8,9]  | Input : [1,2,3,null,null,4,5]    |
| image | image  |

### CODE REVIEW

**Comment by Mrjoy832** [comment]
Can You add an index at top of the readme please?? only taking name of all existing traversal

**Comment by utkarsh006** [comment]
Yes surely I will do this ASAP :)

**Comment by utkarsh006** [comment]
> Can You add an index at top of the readme please?? only taking name of all existing traversal

@Mrjoy832 I have made the requested changes :) Please review it .

**Review by Mrjoy832 — APPROVED** [review]

**Comment by Mrjoy832** [comment]
LGTM🌟

File: T-Tree/T-TreeTraversals/README.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+# CONTENTS
```
Comment (Lakhankumawat) [inline T-Tree/T-TreeTraversals/README.md:1]: `Contents` Follow Camelcasing

File: T-Tree/T-TreeTraversals/README.md [hunk 2]
```diff
@@ -55,3 +60,35 @@ input output 1
 input output 2
 ![Screenshot 2022-03-03 202230](https://user-images.githubusercontent.com/78430607/156589622-279cf956-58e5-4672-9d8e-1187201d38ab.png)
 ![Screenshot 2022-03-03 202257](https://user-images.githubusercontent.com/78430607/156589641-7dfd9e98-c523-4375-b39a-07127f84720b.png)
+
+
+# DIAGONAL TRAVERSAL
```
Comment (Lakhankumawat) [inline T-Tree/T-TreeTraversals/README.md:65]: Came Casing ` Diagonal Traversal`

File: T-Tree/T-TreeTraversals/README.md [hunk 3]
```diff
@@ -55,3 +60,35 @@ input output 1
...
+
+
+# DIAGONAL TRAVERSAL 
+- [PROBLEM STATEMENT](#problem-statement)
+- [Examples](#examples)
+- [Algorithm](#algorithm-1)
+- [code](https://github.com/utkarsh006/LearnCPP/blob/Diagonal_Traversal/T-Tree/T-TreeTraversals/DiagonalTraversal.cpp)
+- [Time Complexity and Space Complexity](#time-complexity-and-space-complexity)
+
+## PROBLEM STATEMENT
```
Comment (Lakhankumawat) [inline T-Tree/T-TreeTraversals/README.md:72]: Camel

File: T-Tree/T-TreeTraversals/README.md [hunk 4]
```diff
@@ -55,3 +60,35 @@ input output 1
...
+- The function will return the elements being placed in diagonal fashion in binary tree.
+- We have used an unordered map for this purpose.
+- By recursion we have solved one case and pushed the curr node in the map.
+- Then in recursive manner we have called the same function for each left and right children of the given binary tree.
+- The hypothesis of Base case lies in the fact that if there is no node then we would be returning NULL.
+- Then we have traversed through the map to print the stored elements.
+
+## Time Complexity and Space Complexity
+- TIME COMPLEXITY : O(N)
+- SPACE COMPLEXITY : O(N)
```
Comment (Lakhankumawat) [inline T-Tree/T-TreeTraversals/README.md:94]: Camel Casing

**Review by Lakhankumawat — CHANGES_REQUESTED** [review]

**Comment by Lakhankumawat** [comment]
I have an offer for you @utkarsh006  do me a favour and i will increase level of your pr 
Visit this https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md and improve the README like this , Like 
1. table of contents at the top 
2. remove output screenshots from morris sort and modify it too a little bit .
3. Add a working image of morris sort above just like you did for your sort .

**Comment by utkarsh006** [comment]
> I have an offer for you @utkarsh006 do me a favour and i will increase level of your pr Visit this https://github.com/Lakhankumawat/LearnCPP/blob/main/S-SortingAlgorithms/README.md and improve the README like this , Like
> 
> 1. table of contents at the top
> 2. remove output screenshots from morris sort and modify it too a little bit .
> 3. Add a working image of morris sort above just like you did for your sort .
I am so sorry I cant perform the above given tasks as I dont know about Morris Traversal of Binary Tree and secondly my college exams are going on. I hope you understand :) Hoping for a positive response.

**Comment by Lakhankumawat** [comment]
Do the reqired changes other than the changes which i offered ( optional ones )

**Comment by utkarsh006** [comment]
> Do the reqired changes other than the changes which i offered ( optional ones )
@Lakhankumawat 
I have done the required camel casing changes !

**Review by Lakhankumawat — APPROVED** [review]
## LGTM 🎉✨

**Comment by Lakhankumawat** [comment]
@all-contributors please add @utkarsh006 code ,

**Pull request merged**